README.md
# Microsoft Application Insights Output Plugin for Logstash

[![GitHub version](https://badge.fury.io/gh/microsoft%2Flogstash-output-application-insights.svg)](https://badge.fury.io/gh/microsoft%2Flogstash-output-application-insights)
[![Gem Version](https://badge.fury.io/rb/logstash-output-application_insights.svg)](https://badge.fury.io/rb/logstash-output-application_insights)

This project is a plugin for [Logstash](https://github.com/elastic/logstash).

# Plugin Features

# Supported Logstash Versions

* Logstash 2.3.2
* Logstash 2.3.4

Note: 

* x64 Ruby for Windows is known to have some compatibility issues.
* the plugin depends on azure-storage that depends on gem nokogiri, which doesn't support Ruby 2.2+ on Windows.





# Setting up

## Install Logstash

  - Download logstash from https://www.elastic.co/downloads/logstash

## Install logstash-output-application_insights output plugin

One command installation:
```sh
bin/logstash-plugin install "logstash-output-application_insights"
```

## Create configuration file

Example (input from files output Application Insights):
```ruby
input {
  file {
    path => "/../files/*"
    start_position => "beginning"
  }
}
filter {
  # some filters here
}
output {
  application_insights {
    instrumentation_key => "5a6714a3-ec7b-4999-ab96-232f1da92059"
    table_id => "c24394e1-f077-420e-8a25-ef6fdf045938"
    storage_account_name_key => [ "my-storage-account", "pfrYTwPgKyYNfKBY2QdF+v5sbgx8/eAQp+FFkGpPBnkMDE1k+ZNK3r3qIPqqw8UsOIUqaF3dXBdPDouGJuxNXQ==" ]
  }
}
```

## Run Logstash
```sh
bin/logstash -f 'file://localhost/../your-config-file'
```





# Installation options

## One command installation:
```sh
bin/logstash-plugin install "logstash-output-application_insights"
```

**If above does not work, or you would like to patch code here is a workaround to install this plugin within your logstash:**

- Check out/clone microsoft/logstash-output-application-insights code from github https://github.com/Microsoft/logstash-output-application-insights

## Option 1: Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the logstash-output-application-insights plugin path:
```ruby
gem "logstash-output-application-insights", :path => "/../logstash-output-application-insights"
```

- Install plugin the plugin from the Logstash home
```sh
bin/logstash-plugin install --no-verify
```

## Option 2:  Run in an installed Logstash

- Build your plugin gem
```sh
gem build logstash-output-application-insights.gemspec
```

- Install the plugin from the Logstash home
```sh
bin/logstash-plugin install "logstash-output-application_insights"
```




# Configuration parameters

### storage_account_name_key
Array of pairs, storage_account_name and an array of acces_keys. No default
At least one pair is required. If not defined, values will be taken (if exist) from Environment Variable: AZURE_STORAGE_ACCOUNT and AZURE_STORAGE_ACCESS_KEY
examples:
```ruby
storage_account_name_key => [ "my-storage-account", "pfrYTwPgKyYNfKBY2QdF+v5sbgx8/eAQp+FFkGpPBnkMDE1k+ZNK3r3qIPqqw8UsOIUqaF3dXBdPDouGJuxNXQ==" ]

storage_account_name_key => [ ["my-storage-account1", "key1"], "my-storage-account2", "key2"], ["my-storage-account3", "key3"] ]

storage_account_name_key => [ ["my-storage-account1", ["key11", "key12"]], ["my-storage-account1", "key2"], ["my-storage-account1", ["key3"] ]
```

### azure_storage_table_prefix
A prefix for the azure storage tables name used by this Logstash instance. Default host name
It is recommeded that each Logstash instance have a unique prefix, to avoid confusion and loss of tracking, although sharing tables won't damage proper execution.
If not set, the host name is used (not alphanumeric characters are removed, and converted downcase), if host name available.
The prefix string may contain only alphanumeric characters, it is case sensitive, and must start with a letter 
example:
```ruby
azure_storage_table_prefix => "myprefix"
```

### azure_storage_container_prefix
A prefix for the azure storage containers name used by this Logstash instance. Default host name
It is recommeded that each Logstash prefix have a unique prefix, to avoid confusion and loss of tracking, although sharing containers won't damage proper execution.
if not set, the host name is used (not alphanumeric characters are removed, and converted downcase), if host name available.
The prefix string may contain only alphanumeric characters and dash, double dash are not allowed, it is case insesitive.
example:
```ruby
azure_storage_container_prefix => "myprefix"
```

### azure_storage_blob_prefix
A prefix for the azure storage blobs name used by this Logstash instance. Default host name
Each Logstash prefix MUST have a unique prefix, to avoid loss of data !!!
If not set, the host name is used (not alphanumeric characters are removed, and converted downcase), if host name available
string may include only characters that are allowed in any valid url
example:
```ruby
azure_storage_blob_prefix => "myprefix"
```

### instrumentation_key
Default Application Insights Analytics instrumentation_key. No default
It will be used only in case the key is not specified in the tables property associated to a table_id, or as field or metadata fields in the event
example:
```ruby
instrumentation_key => "5A6714A3-EC7B-4999-AB96-232F1DA92059"
```

### table_id
Default Application Insights Analytics table_id. No default
Will be used only in case it is not specified as field or metadata fields in the event
example:
```ruby
table_id => "C24394E1-F077-420E-8A25-EF6FDF045938"
```

### table_columns
Specifies the list of the fields that will be filtered from the event, fields not specified will be ignored. No Default (event all fields)
If not specified all fileds in events will be filtered, the order is kept. The order is essential in case of CSV serialization.
example:
```ruby
table_columns => [ "EventLogID", "AppName", "EnvironmentName", "ActivityID", "EventID", "Severity", "Title" ]
```

### case_insensitive_columns
If set to true, events fields are refered as case insensitive. Default false (case sensitive)
example:
```ruby
case_insensitive_columns => true
```

### blob_max_bytesize
Advanced, internal, should not be set. Default 1 GB.
Azure storage maximum bytesize is 192 GB ( = 50,000 * 4 MB ) 
example:
```ruby
blob_max_bytesize => 4000000000
```

### blob_max_events
Specifies, maximum number of events in one blob. Default 256,000 events
Setting it too low may improve latency, but will reduce ingestion performance
Setting it too high may damage latency up to maximum delay, but ingestion will be more efficient, and load on network will be lower
example:
```ruby
blob_max_events => 1000000
```

### blob_max_delay
Specifies maximum latency time, in seconds. Defualt 60 seconds
The latency time is measured since the time an event arrived till it is commited to azure storage, and Application Insights is notified.
The total latency time may be higher, as this is not the full ingestion flow 
example:
```ruby
blob_max_delay => 3600
```

### blob_serialization
Specifies the blob serialziation to create. Default "json"
currently 2 types are supported "csv" and "json""
example:
```ruby
blob_serialization => "json""
```

### io_retry_delay
Interval of time between retries due to IO failures
example:
```ruby
io_retry_delay => 0.5
```

### io_max_retries
Number of retries on IO failures, before giving up, and move to available options
example:
```ruby
io_max_retries => 3
```

### blob_retention_time
Specifies the retention time of the blob in the container after it is notified to Application Insighta Analytics. Dfeauly 604,800 seconds (1 week)
Once the retention time expires, the blob is the deleted from container
example:
```ruby
blob_retention_time => 604800
```

### blob_access_expiry_time
Specifies the time Application Insights Analytics have access to the blob that are notifie. Default 86,400 seconds ( 1 day)
Blob access is limited with SAS URL
example:
```ruby
blob_retention_time => 604800
```

### csv_default_value
Specifies the string that is used as the value in a csv record, in case the field does not exist in the event. Default ""
example:
```ruby
csv_default_value => "-"
```

### serialized_event_field
Specifies a serialized event field name, that if exist in current event, its value as is will be taken as the serialized event. No Default
example:
```ruby
serialized_event_field => "serializedMessage"
```


### logger_level
Specifies the log level. valid values are: DEBUG, INFO, WARN, ERROR, FATAL, UNKNOWN. Default "INFO"
example:
```ruby
logger_level => "INFO"
```

### logger_files
Specifies the list of targets for the log. may include files, devices, "stdout" and "stderr". Default "logstash-output-application-insights.log"
example:
```ruby
csv_default_value => [ "c:/logstash/dev/runtime/log/logstash-output-application-insights.log", "stdout" ]
```

### logger_progname
Specifies the program name that will displayed in each log record. Default "AI"
Should be modified only in case there is another plugin with the same program name
example:
```ruby
logger_progname => "MSAI"
```

### logger_shift_size
Specifies maximum logfile size. No Default (no size limit)
Only applies when shift age is a number !!!
Not supported in Windows !!!
example (1 MB):
```ruby
logger_shift_size => 1048576
```

### logger_shift_age
Specifies Number of old logfiles to keep, or frequency of rotation (daily, weekly or monthly). No default (never)
Not supported in Windows !!!
examples:
```ruby
logger_shift_age => weekly
```

```ruby
logger_shift_age => 5
```

### resurrect_delay
Specifies the time interval, between tests that check whether a stoarge account came back to life, after it stoped responding. Default 10 seconds
example (half second):
```ruby
flow_control_delay => 0.5
```

### flow_control_suspend_bytes
Specifies the high water mark for the flow control, that is used to avoid out of memory crash. Default 52,428,800 Bytes (50 MB)
Once the memory consumption reach the high water mark, the plugin will stop accepting events, till memory is below the low water mark
example (200 MB):
```ruby
flow_control_suspend_bytes => 209715200
```

### flow_control_resume_bytes
Specifies the low water mark for the flow control, that is used to avoid out of memory crash. Default 41,820,160 Bytes (40 MB)
Once memory consumption reach the high water mark, the plugin will stop accepting events, till memory is below the low water mark
example (10 MB):
```ruby
flow_control_resume_bytes => 10455040
```

### flow_control_delay
Specifies the amount of time the flow control suspend receiving event. Default 1 second
It is to allow GC, and flush of event to Azure storage before checking whether memory is below low water mark
example (half second):
```ruby
flow_control_delay => 0.5
```

### ca_file
File path of the CA file, required only if having issue with SSL (see OpenSSL). No default
example:
```ruby
ca_file => "/path/to/cafile.crt"
```

### disable_telemetry
When set to true, telemetry about the plugin, won't be sent to Application Insights. Deafult false
example:
```ruby
disable_telemetry => true
```

### disable_cleanup
When set to true, storage cleanup won't be done by the plugin (should be done by some other means or by another Logstash process with this flag enabled) Default false
example:
```ruby
disable_cleanup => true
```

### delete_not_notified_blobs
When set to true, not notified blobs are deleted, if not set they are copied to the orphan-blobs container. Default false
example:
```ruby
delete_not_notified_blobs => true
```
### validate_endpoint
When set to true, access to application insights will be validated at initialization and if validation fail, logstash process will abort. Default false
example:
```ruby
validate_endpoint => true
```

### validate_storage
When set to true, access to azure storage for each of the configured accounts will be validated at initialization and if validation fail, logstash process will abort. Default false
example:
```ruby
validate_storage => true
```

### save_notified_blobs_records
When set to true, notified blobs records are saved in the state table, as long as blobs are retained in their containers. Default false
Used for troubleshooting
example:
```ruby
save_notified_blobs_records => true
```

### disable_notification
When set to true, notification is not sent to application insights, but behaves as if notified. Default false
Used for troubleshooting
example:
```ruby
disable_notification => true
```

### disable_blob_upload
When set to true, events are not uploaded, and blob not commited, but behaves as if uploaded and uploaded. Default false
Used for troubleshooting
example:
```ruby
disable_blob_upload => true
```

### stop_on_unknown_io_errors
When set to true, process will stop if an unknown IO error is detected. Default false
Used for troubleshooting
example:
```ruby
stop_on_unknown_io_errors => true
```

### azure_storage_host_suffix
when set an alternative storage service will be used. Default "core.windows.net"
example:
```ruby
azure_storage_host_suffix => "core.windows.net"
```

### application_insights_endpoint
when set blob ready notification are sent to an alternative endpoint. Default "https://dc.services.visualstudio.com/v2/track"
example:
```ruby
application_insights_endpoint => "https://dc.services.visualstudio.com/v2/track"
```

### notification_version
Advanced, internal, should not be set, the only current valid value is 1
example:
```ruby
notification_version => 1
```

### tables
Allow to support multiple tables, and to configure each table with its own parameters, using the global parameters as defaults.
It is only required if the plugin need to support mutiple table.
Tables is Hash, where the key is the table_id and the value is a has of specific properties, that their defualt value are the global properties.
The specific properties are: instrumentation_key, table_columns, blob_max_delay, csv_default_value, serialized_event_field, blob_serialization, csv_separator
template:
```ruby
tables => { "table_id1" => { properties } "table_id2" => { properties } }
```

Examples:
```ruby
tables => { "6f29a89e-1385-4317-85af-3ac1cea48058" => { "instrumentation_key" => "76c3b8e9-dfc6-4afd-8d4c-3b02fdadb19f", "blob_max_delay" => 60 } }
```

```ruby
tables => { "6f29a89e-1385-4317-85af-3ac1cea48058" => { "instrumentation_key" => "76c3b8e9-dfc6-4afd-8d4c-3b02fdadb19f", "blob_max_delay" => 60 }
            "2e1b46aa-56d2-4e13-a742-d0db516d66fc" => { "instrumentation_key" => "76c3b8e9-dfc6-4afd-8d4c-3b02fdadb19f", "blob_max_delay" => 120 "ext" => "csv" "serialized_event_field" => "message" } 
          }
```


# Enviroment variables

###AZURE_STORAGE_ACCOUNT
Specifies the Azure storage account name
Will be used by the plugin to set the account name part in plugin property **storage_account_name_key** if it is missing
Example:
```sh
AZURE_STORAGE_ACCOUNT="my-storage-account"
```

###AZURE_STORAGE_ACCESS_KEY
Specifies the Azure storage account access key
Will be used by the plugin to set the key part in plugin property **storage_account_name_key** if it is missing
Example:
```sh
AZURE_STORAGE_ACCESS_KEY="pfrYTwPgKyYNfKBY2QdF+v5sbgx8/eAQp+FFkGpPBnkMDE1k+ZNK3r3qIPqqw8UsOIUqaF3dXBdPDouGJuxNXQ=="
```

# Setting up Http/Https Proxy

If you use a proxy server or firewall, you may need to set the HTTP_PROXY and/or HTTPS_PROXY **environment variables** in order to access Azure storage and Application Insights.
Examples:
```sh
HTTP_PROXY=http://proxy.example.org
HTTPS_PROXY=https://proxy.example.org
```

- If the proxy server requires a user name and password, include them in the following form:
```sh
HTTP_PROXY=http://username:password@proxy.example.org
```

- If the proxy server uses a port other than 80, include the port number:
```sh
HTTP_PROXY=http://username:password@proxy.example.org:8080
```

# Setting up SSL certificates

When using SSL/HTTPS, typically log in or authentication may require a CA Authority (CA) certificate. If the required certificate is not already bundled in the system. it may be configured in the plugin (see above ca_file)
example:
```ruby
ca_file => "/path/to/cafile.crt"
```

# Getting Started for Contributors

If you would like to become an active contributor to this project please follow the instructions provided in [CONTRIBUTING.md](CONTRIBUTING.md) and  [DEVELOPER.md](DEVELOPER.md)

# Provide Feedback

If you encounter any bugs with the library please file an issue in the [Issues](https://github.com/Microsoft/logstash-output-application-insights/issues) section of the project.

# Code of Conduct 

This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

CHANGELOG.md
## 0.1.1 (Aug 22, 2016)

* Plugin published




lib/logstash/outputs/msai.rb
# encoding: utf-8
#-------------------------------------------------------------------------
# # Copyright (c) Microsoft and contributors. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#--------------------------------------------------------------------------

require "logstash/outputs/base"
require "logstash/namespace"
require 'logstash-core/version'

require "stud/interval"

require "azure/storage"
require 'azure/storage/core/auth/shared_access_signature'

require 'azure/core/http/http_request'
require 'azure/core/http/http_response'

require 'rbconfig'
require 'faraday'
require 'openssl'
require "time"
require "json"
require "uri"
require "socket"

require "thread"
require "concurrent" # for atomic and thread safe operations
require "logger"
require "csv"

require "application_insights"

# An msai output that does nothing.
class LogStash::Outputs::Msai < LogStash::Outputs::Base
  require "logstash/outputs/msai/version"
  require "logstash/outputs/msai/utils" 
  require "logstash/outputs/msai/constants" 
  require "logstash/outputs/msai/config" 
  require "logstash/outputs/msai/blob" 
  autoload :Block, "logstash/outputs/msai/block"
  autoload :Storage_cleanup, "logstash/outputs/msai/storage_cleanup"

  autoload :Clients, "logstash/outputs/msai/clients" 
  autoload :Client, "logstash/outputs/msai/client" 

  autoload :Channels, "logstash/outputs/msai/channels"
  autoload :Channel, "logstash/outputs/msai/channel"
  autoload :Sub_channel, "logstash/outputs/msai/sub_channel" 
  autoload :State, "logstash/outputs/msai/state"
  autoload :Flow_control, "logstash/outputs/msai/flow_control"
  autoload :Shutdown, "logstash/outputs/msai/shutdown"
  autoload :Telemetry, "logstash/outputs/msai/telemetry"

  require "logstash/outputs/msai/notify" 
  require "logstash/outputs/msai/exceptions" 
  require "logstash/outputs/msai/timer" 
  require "logstash/outputs/msai/multi_io_logger" 

  config_name "msai"
    
  # default :codec, "json_lines" 
  
  # Array of pairs, storage_account_name and an array of acces_keys 
  # examples [ account1, key1 ]
  #          [ [account1, key1], [accoun2, key2], [account3, key3] ]
  #          [ [account1, [key11, key12]], [account2, key2], [account3, [key3] ]
  config :storage_account_name_key, :validate => :array, :required => true

  # prefix for the azure storage tables used by this Logstash instance
  # it is recommeded that each Logstash prefix have a unique prefix, to avoid
  # confusion and loss of tracking, although sharing tables won't damage proper execution
  # if not set, the machine name is used (not alphanumeric characters are removed, and converted downcase), if machine name available
  # string may contain only alphanumeric character, case sensitive, and must start with a letter 
  config :table_prefix, :validate => :string

  # prefix for the azure storage containers used by this Logstash instance
  # it is recommeded that each Logstash prefix have a unique prefix, to avoid
  # confusion and loss of tracking, although sharing containers won't damage proper execution
  # if not set, the machine name is used (not alphanumeric characters are removed, and converted downcase), if machine name available
  # string may contain only alphanumeric character and dash, double dash not allowed, case insesitive
  config :container_prefix, :validate => :string

  # prefix for the azure storage blobs used by this Logstash instance
  # Each Logstash prefix MUST have a unique prefix, to avoid
  # loss of data !!!
  # if not set, the machine name is used (not alphanumeric characters are removed, and converted downcase), if machine name available
  # string may include only characters that are allowed in any valid url
  config :blob_prefix, :validate => :string
  
  # Default Application Insights Analytics intrumentation_key
  # will be used only in case it is not specified as a table_id property in table_ids_properties
  # or as part of the event's fields or event's metadata fields
  config :intrumentation_key, :validate => :string

  # Default Application Insights Analytics table_id
  # will be used only in case it is not specified as part o
  # of the event's fields or event's metadata fields
  config :table_id, :validate => :string

  # A hash of table_ids, where each table_id points to a set of properties
  # the properties are a hash, where the keys are are the properties
  # current supported properties per table_id are:
  # intrumentation_key, ext, fields_map, csv_default_value, csv_separator, max_delay, event_separator, data_field
  # intrumentation_key, Application Insights Analytics intrumentation_key, will be used in case not specified in any of the event's fields or events's metadata fileds
  # data_field, specifies the data field that may contain the full serialized event (either as json or csv), 
  #             when specified, the ext property should be set either to csv or to json (json is the default)
  #             if event. does not conatin the field, value will be created based on the fileds in the evnt, according to fields_map if configured, or all fileds in event
  #             if event contains this filed, and ext is csv
  #                  if value is a string, it will be used as is as the serialized event, without validating whether it is a csv string
  #                  if value is an array, it will be serialized as an array of csv columns
  #                  if value is a hash, it will be serialized based on fields_map to csv columns
  #             if event contains this filed, and ext is json
  #                  if value is a string, it will be used as is as the serialized event, without validating whether it is a json string
  #                  if value is a hash, it will be serialized to json, if fileds_map exit, it will be based on filds_map
  #                  if value is an array, it will be zipped with fields_map (if exist) and serialized to json
  # ext, blob extension, the only valid values are either csv or json, 
  #             should be set whenever the default json is not appropriate (.e, csv)
  # max_delay, maximum latency time, in seconds, since the time the event arrived till it should be commited in azure storage, and Application Insights is notified
  # event_separator, specifies the string that is used as a separator between events in the blob
  # fields_map, specifies the event fields that should be serialized, and their order (order is required for csv)
  #          if csv serialization will be used for this table_id
  #          each fields_map field is a hash with 3 keys: name, type, and default. Only name is mandatory
  #               name - is the name of the event fleld that its value should be mapped to this columns
  #               type - is the type of this field: "string", "hash", "array", "number", "json", "boolean", "float", "integer", "dynamic", "datetime", "object"
  #               default - is the value to be used for this column, in case the field is missing in the event
  # csv_separator, specifies the string that is used as a separator between columns, 
  #                can be specified only together with fields_map
  # csv_default_value, specifies the string that is used as the value in a csv record, in case the field does not exist in the event
  #                can be specified only together with fields_map
  #
  # Example json table_id
  #   table_ids_properties => {"a679fbd2-702c-4c46-8548-80082c66ef28" => {"intrumentation_key" => "abee940b-e648-4242-b6b3-f2826667bf96", "max_delay" => 60} }
  # Example json table_id, input in data_field
  #   {"ab6a3584-aef0-4a82-8725-2f2336e59f3e" => {"data_field" => "message". "ext" => "json"} }
  # Example csv table_id, input in data_field
  #   {"ab6a3584-aef0-4a82-8725-2f2336e59f3e" => {"data_field" => "csv_message". "ext" => "csv"} }
  # Example csv table_id, input in event fields
  #   {"ab6a3584-aef0-4a82-8725-2f2336e59f3e" => { "ext" => "csv", "fields_map" => [ {name => "Timestamp" type => datetime }, "Value", "Custom" ] } }
  # Example csv table_id, input in event fields
  #   {"ab6a3584-aef0-4a82-8725-2f2336e59f3e" => { "ext" => "json", "fields_map" => [ "Timestamp", "Value", "Custom" ] } }

  config :table_ids_properties, :validate => :hash, :default => {}

  # Advanced, internal, should not be set, the default is 192 GB ( = 50,000 * 4 MB ) 
  # azure storage maximum number of blocks per blob is 192 GB ( = 50,000 * 4 MB ) 
  config :blob_max_bytesize, :validate => :number

  # Specifies, maximum number of events in one blob,
  # setting it low may reduce the latency, but reduce logstash performance
  # setting it high may increase latency to maximum delay, but logstash will be more efficient, and load on network will be lower
  config :blob_max_events, :validate => :number

  # Specifies maximum latency time, in seconds, since the time the event arrived 
  # till it is commited to azure storage, and Application Insights is notified
  # The total latency may be higher, as this is not the full ingestion flow 
  config :blob_max_delay, :validate => :number

  # Interval of time between retries due to IO failures
  config :io_retry_delay, :validate => :number

  # Number of retries on IO failures, before giving up, and move to available options
  config :io_max_retries, :validate => :number

  # Specifies the retention time of the blob in the container after it is notified to Application Insighta Analytics
  # Once the retention time expires, the blob is the deleted from container
  config :blob_retention_time, :validate => :number

  # Specifies the time Application Insights Analytics have access to the blob that are notifie
  # Blob access is limited with SAS URL
  config :blob_access_expiry_time, :validate => :number

  # Advanced, internal, should not be set, the default is \r\n, 
  # specifies the string that is used as a separator between events in the blob
  config :event_separator, :validate => :string

  # Advanced, internal, should not be set, the default is comma, 
  # specifies the string that is used as a separator between columns in a csv record
  config :csv_separator, :validate => :string

  # specifies the string that is used as the value in a csv record, in case the field does not exist in the event, the default is ""
  config :csv_default_value, :validate => :string

  # specifies the log level. valid values are: DEBUG, INFO, WARN, ERROR, FATAL, UNKNOWN
  config :logger_level, :validate => :string

  # Specifies the list of targets for the log. may include files, devices, "stdout: and "stderr"
  config :logger_files, :validate => :array

  # Advanced, internal, should not be set, the default is MSAI, 
  # Specifies the program name that will displayed in each log record
  config :logger_shift_progname, :validate => :string

  # Specifies when file logs are shifted. valid values are either an integer or "daily", "weekly" or "monthly"
  config :logger_shift_size

  # Specifies the shift age of a log
  config :logger_shift_age, :validate => :number

  # Specifies the time interval, between tests that check whether a stoarge account came back to life, 
  # after it stoped responding 
  config :resurrect_delay, :validate => :number

  # specifies the high water mark for the flow control, that used to avoid out of memory crash
  # once memory consumption reach reach the high water mark, the plugin will stop accepting events, till memory
  # is below the low water mark
  config :flow_control_suspend_bytes, :validate => :number

  # specifies the low water mark for the flow control, that used to avoid out of memory crash
  # once memory consumption reach the high water mark, the plugin will stop accepting events, till memory
  # is below the low water mark
  config :flow_control_resume_bytes, :validate => :number

  # specifies the amount of time the flow control suspend receiving event, to allow GC, and flush of event to Azure storage
  # before checking whether memory is below low water mark
  config :flow_control_delay, :validate => :number

  # File path of the CA file if having issue with SSL
  config :ca_file, :validate => :string


  # When set to true, telemetry about the plugin, won't be sent to Application Insights
  config :disable_telemetry, :validate => :boolean

  # When set to true, storage cleanup won't be done by the plugin (should be done by some other means or by another Logstash process with this flag enabled)
  config :disable_cleanup, :validate => :boolean


  # Advanced, internal, should not be set, the default is false
  # When set to true, notification is not sent to application insights, but behaves as if notified
  config :disable_notification, :validate => :boolean

  # Advanced, internal, should not be set, the default is false
  # When set to true, events are not uploaded, and blob not commited, but behaves as if uploaded and notified
  config :disable_blob_upload, :validate => :boolean

  # Advanced, internal, should not be set, the default is false
  # When set to true, process will stop if an unknown IO error is found
  config :stop_on_unknown_io_errors, :validate => :boolean

  # Advanced, internal, should not be set, the default is Application Insights production endpoint
  # when set notification are sent to an alternative endpoint, used for internal testing
  config :notification_endpoint, :validate => :string

  # Advanced, internal, should not be set, the only current valid value is 1
  config :notification_version, :validate => :number

  public

  def register

    # set configuration
    Config.validate_and_adjust_configuration( default_configuration )
    configuration = Config.current

    Multi_io_logger.config( configuration )

    # be careful don't use here @logger, as it will override Logstash @logger, and may create starnge behaviour
    @private_logger = configuration[:logger]

    @private_logger.info { "configuration: #{configuration}" }

    @telemetry = Telemetry.instance
    configuration[:telemetry_channel] = @telemetry.telemetry_channel

    Timer.config( configuration )
    Blob.config( configuration )


    @shutdown = Shutdown.instance
    @channels = Channels.instance

    @storage_cleanup = Storage_cleanup.start

    @private_logger.info { "plugin registered" }

    # @codec.on_event do |event, encoded_event|
    #   @channels.receive( event, encoded_event )
    # end

    return "ok\n"
  end # def register
  

  def receive ( event )
    # @codec.encode( event )
    @channels.receive( event, nil )
    return "ok\n"
  end

  def close
    @shutdown.submit
  end

  private

  # -----------------------------------------------


  def list_blob_names
    blob_names = Set.new []
    loop do
      continuation_token = NIL
      entries = @azure_blob.list_blobs(@container, { :timeout => 10, :marker => continuation_token})
      @@logger.debug { 'blob entries: #{entries}' }
      entries.each do |entry|
        @@logger.debug { 'blob entry name: #{entry.name}' }
        blob_names << entry.name
      end
      continuation_token = entries.continuation_token
      break if continuation_token.empty?
    end
    return blob_names
  end # def list_blobs


  def list_container_names
    container_names = Set.new []
    loop do
      continuation_token = NIL
      containers = @azure_blob.list_containers()
      @@logger.debug { 'containers: #{containers}' }
      containers.each do |container|
        @@logger.debug { 'container entry name:' + container.name }
        container_names << container.name
        upload(container.name, "blob-append-" + container.name, "test - " + container.name)
        blobs = @azure_blob.list_blobs(container.name)
        blobs.each do |blob|
          @@logger.debug { 'blob name: ' + blob.name }
        end
      end
      continuation_token = containers.continuation_token
      break if continuation_token.empty?
    end
    return container_names
  end # def list_blobs

  def create_container (container_name)
    begin
      @azure_blob.create_container(container_name)
    rescue
      @@logger.debug { $! }
    end
  end
end # class LogStash::Outputs::Msai

